drivers/can/can_sja1000.c

```diff
@@ -0,0 +1,707 @@
+#define CAN_SJA1000_RESET_MODE_DELAY	    \
+	K_USEC(CAN_SJA1000_RESET_MODE_TIMEOUT_USEC / CAN_SJA1000_RESET_MODE_RETRIES)
+
+static inline void can_sja1000_write_reg(const struct device *dev, uint8_t reg, uint8_t val)
+{
+	const struct can_sja1000_config *config = dev->config;
+
+	LOG_DBG("write reg %d = 0x%02x", reg, val);
+	return config->write_reg(dev, reg, val);
+}
+
+static inline uint8_t can_sja1000_read_reg(const struct device *dev, uint8_t reg)
+{
+	const struct can_sja1000_config *config = dev->config;
+	uint8_t val;
+
+	val = config->read_reg(dev, reg);
+	LOG_DBG("read reg %d = 0x%02x", reg, val);
+
+	return val;
+}
+
+static inline int can_sja1000_enter_reset_mode(const struct device *dev)
+{
+	int retries = CAN_SJA1000_RESET_MODE_RETRIES;
+	uint8_t mod;
+
+	mod = can_sja1000_read_reg(dev, CAN_SJA1000_MOD);
+
+	while ((mod & CAN_SJA1000_MOD_RM) == 0) {
+		if (--retries < 0) {
+			return -EIO;
+		}
+
+		can_sja1000_write_reg(dev, CAN_SJA1000_MOD, mod | CAN_SJA1000_MOD_RM);
+		k_sleep(CAN_SJA1000_RESET_MODE_DELAY);
+		mod = can_sja1000_read_reg(dev, CAN_SJA1000_MOD);
+	};
+
+	return 0;
+}
+
+static inline int can_sja1000_leave_reset_mode(const struct device *dev)
+{
+	int retries = CAN_SJA1000_RESET_MODE_RETRIES;
+	uint8_t mod;
+
+	mod = can_sja1000_read_reg(dev, CAN_SJA1000_MOD);
+
+	while ((mod & CAN_SJA1000_MOD_RM) == 1) {
+		if (--retries < 0) {
+			return -EIO;
+		}
+
+		can_sja1000_write_reg(dev, CAN_SJA1000_MOD, mod & ~(CAN_SJA1000_MOD_RM));
+		k_sleep(CAN_SJA1000_RESET_MODE_DELAY);
+		mod = can_sja1000_read_reg(dev, CAN_SJA1000_MOD);
+	};
+
+	return 0;
+}
+
+int can_sja1000_set_timing(const struct device *dev, const struct can_timing *timing)
+{
+	const struct can_sja1000_config *config = dev->config;
+	struct can_sja1000_data *data = dev->data;
+	uint8_t btr0;
+	uint8_t btr1;
+	uint8_t sjw;
+	int err;
+
+	__ASSERT_NO_MSG(timing->sjw == CAN_SJW_NO_CHANGE || (timing->sjw >= 1 && timing->sjw <= 4));
+	__ASSERT_NO_MSG(timing->prop_seg == 0);
+	__ASSERT_NO_MSG(timing->phase_seg1 >= 1 && timing->phase_seg1 <= 16);
+	__ASSERT_NO_MSG(timing->phase_seg2 >= 1 && timing->phase_seg2 <= 8);
+	__ASSERT_NO_MSG(timing->prescaler >= 1 && timing->prescaler <= 64);
+
+	k_mutex_lock(&data->mod_lock, K_FOREVER);
+
+	err = can_sja1000_enter_reset_mode(dev);
+	if (err != 0) {
+		goto unlock;
+	}
+
+	if (timing->sjw == CAN_SJW_NO_CHANGE) {
+		sjw = data->sjw;
+	} else {
+		sjw = timing->sjw;
+		data->sjw = timing->sjw;
+	}
+
+	btr0 = CAN_SJA1000_BTR0_BRP_PREP(timing->prescaler - 1) |
+	       CAN_SJA1000_BTR0_SJW_PREP(sjw - 1);
+	btr1 = CAN_SJA1000_BTR1_TSEG1_PREP(timing->phase_seg1 - 1) |
+	       CAN_SJA1000_BTR1_TSEG2_PREP(timing->phase_seg2 - 1);
+
+	if ((data->mode & CAN_MODE_3_SAMPLES) != 0) {
+		btr1 |= CAN_SJA1000_BTR1_SAM;
+	}
+
+	can_sja1000_write_reg(dev, CAN_SJA1000_BTR0, btr0);
+	can_sja1000_write_reg(dev, CAN_SJA1000_BTR1, btr1);
+
+	err = can_sja1000_leave_reset_mode(dev);
+	if (err != 0) {
+		goto unlock;
+	}
```

Comment:
Is it necessary?

Reply:
It seems okay, no? The mutex in this place is already held

Reply:
Ahh now I get it, the unlock is right below.

Well for consistency I would keep, the compiler is smart enough to optimize here.